### PR TITLE
UCP/TAG: Fix sync send for UCT zcopy message sizes

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

Fixes bug in `ucp_tag_send_sync_nb` for UCT zcopy message sizes, when `uct_ep_am_zcopy` returns `UCS_OK`

## Why ?

`ucp_tag_send_sync_nb` doesn't work for UCT zcopy message sizes, when `uct_ep_am_zcopy` returns `UCS_OK`, i.e. the request is never completed

## How ?

`ucp_tag_send_nb`:
- calls `ucp_proto_am_zcopy_req_complete` to complete Zcopy request if the operation completes immediately (i.e `uct_ep_am_zcopy` returns `UCS_OK`).
- calls `ucp_proto_am_zcopy_completion` in Zcopy completion callback with checks for `req->send.state.dt.offset == req->send.length` when Zcopy operation completes not immediately (i.e. `uct_ep_am_zcopy` returns `UCS_INPROGRESS`)

Apply the same approach for `ucp_tag_send_sync_nb` as we do for `ucp_tag_send_nb`:
- calls `ucp_tag_eager_sync_zcopy_req_complete`, if `uct_ep_am_zcopy` returns `UCS_OK`
- calls `ucp_tag_eager_sync_zcopy_completion`, if `uct_ep_am_zcopy` returns `UCS_INPROGRESS` in Zcopy completion callback

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10820/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 33 workers (click for details)</summary>

**Note:** the logs will be deleted after 02-Jul-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/1491cbe0fb2c9785b9af3f8c94cd6e23be4699b2/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/45200c93f6287493bfc773bd892e9d02f1a23b70/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/291e1eb522340215af22ccb894ea1369be8a6a01/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/fa35477a4bb5e4207f9b240fbc084be4fe30c6ea/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/912c38333f2444c171338ad2d68e9e7547e96182/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/82d26a681c7357029ee54f6677d9a185874d2c97/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/583d9833d84205b90ec68b8c69e84d4155f83e2e/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/bc5360552efe60eac1fcaffa136e79de7eec59b4/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/21114b9fdb4eacd6d258c582b60f8af8f244a6c8/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W0](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/efc27c846e62d688b500fdf7a439e521b220c2f3/hpc-test-node-upstream_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W1](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/698838fe4a7e55aec19f2e9134811bf53f7682ad/hpc-test-node-upstream_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W2](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/beb1999940c341dafaca1b9bfee5f28b43e7c84b/hpc-test-node-upstream_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W3](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/584c82f27bad31d296e76a0970716fc07e027449/hpc-test-node-upstream_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W0](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/92519bce83a6a2baa30eeedb9402d7477d4308f7/hpc-test-node2_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W1](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/2fda1518314cb1850591a99e10aff8fc41b6974b/hpc-test-node2_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W2](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/0695ba594fd0eaa60c6dca2ac8e3906e224ba3fe/hpc-test-node2_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W3](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/8ddc22bd13b69eacd5b9f0d032e5b84f040cfc81/hpc-test-node2_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W0](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/a29cee0b317f3e5a9bc9682fbbf7c714e4c6f077/hpc-test-node3_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W1](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/c787d453fcf7afa8eb7bf23d04c3189966b2e043/hpc-test-node3_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W2](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/2b34a8b7f246de5b00309b19a6ae0848c151e7c8/hpc-test-node3_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W3](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/df372eb8b4ff6335677c268d9e0ecc6d90966e57/hpc-test-node3_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node_W0](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/5ed63dd4d54f66835ac623137064440b5f9e5ecb/hpc-test-node_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node_W1](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/04ffd93d77fc8c18f56491d9410af79b41848afe/hpc-test-node_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node_W2](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/e70942dd9bfadab6e9cb9b9ebd4cf07f1e5751bb/hpc-test-node_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node_W3](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/5aa8ab4787b4cb2aba2ddf0c374e135b8dff40f5/hpc-test-node_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/e212726e4b57a5e48af42e58688b873c4ded7ec1/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/6502ca81c640648dded9bea28a225670a495f516/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/ce14626e41f98464304de6d20994f18edca88b22/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/a3a75da80182cd835e807d3732d59b7d25ff456e/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/ef9567739d8dd4488bf51d8e76196cd9c4cfd2d9/vulcan-jenkins_W0) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/c9573c695ba15ff0b0eb1e8cc93e92c771de5feb/vulcan-jenkins_W1) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/21138e6beb801d6f9eabd06017266519bee81c14/vulcan-jenkins_W2) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/5d451a32f37dc2fbda7c3dfbb65dd489/raw/a0242a088ddd9bfa1d85393171a92a62471b8058/vulcan-jenkins_W3) | :heavy_check_mark: SUCCESS

**Comment by dmitrygladkov** [comment]
> why is this needed?

@brminich this was an incorrect (or unnecessarily complicated) approach
pls, take a look, this should be better

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10826/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
<details><summary>Mellanox CI: <b>PASSED</b#  > on 33 workers (click for details)</summary>

**Note:** the logs will be deleted after 03-Jul-2019

Agent/Stage | Status
-- | --
[_main](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/ce250b554fc07a3976c7cfd3e2b20c9a304e2645/_main) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/c9dc2101010299fd6150bcf89c1eb0e50c568400/hpc-arm-cavium-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/9b906fd0a2b34fda49aa56578ca42c12f5db98ac/hpc-arm-cavium-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/7e7b286401ae5f93a0749e6328571adf9f5e4ed9/hpc-arm-cavium-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-cavium-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/12cb4ed62f57e63d41ea14c3a17dd14a0ba5d34a/hpc-arm-cavium-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/2b46463a7ce6712a96c50a90b0c3799f5fc7d178/hpc-arm-hwi-jenkins_W0) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/f6456304cbe89820a788c9f5cca0a8038d9a5a57/hpc-arm-hwi-jenkins_W1) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/a6d3c01320d76237a37a3d800266adf06ed13aad/hpc-arm-hwi-jenkins_W2) | :heavy_check_mark: SUCCESS
[hpc-arm-hwi-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/e76ab5576f3385a26c7bb5f891a013491313b897/hpc-arm-hwi-jenkins_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W0](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/d77209bac9d52484ef942a471de52980bde0878f/hpc-test-node-upstream_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W1](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/11ddf25db1f9b60b3c414fe0de314bc7e9771968/hpc-test-node-upstream_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W2](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/6f50852b2d0d5216770f3111eca5ffdd53e8ea09/hpc-test-node-upstream_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node-upstream_W3](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/ce9cd37e1dbcf1d695da2aca259897c6bfc822b7/hpc-test-node-upstream_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W0](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/41ac274bf4d98cc3baef765b9d54db92db00adfc/hpc-test-node2_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W1](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/9be80ffb818e7cf4889b9811248c75e57c35909f/hpc-test-node2_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W2](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/74e1f48995db0f1fd464d6027a86aa2f13bfebd9/hpc-test-node2_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node2_W3](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/b233155e0cfea5560dddcd6f7f5b824e1c5ca824/hpc-test-node2_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W0](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/0a1eaa504757dac9a714646a53b871a1a2eb986b/hpc-test-node3_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W1](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/b7619835a053467e682ad3610da5303babf76bf5/hpc-test-node3_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W2](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/6525122139828d33206655a8a242fc585e3eb8e7/hpc-test-node3_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node3_W3](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/5bb82512cbe4cb33461b7c357af51bfcf4ecb363/hpc-test-node3_W3) | :heavy_check_mark: SUCCESS
[hpc-test-node_W0](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/1b28abc4ea1a686d31d59333c59b1490863c847b/hpc-test-node_W0) | :heavy_check_mark: SUCCESS
[hpc-test-node_W1](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/a6d7722274ab7513c5ec11e9d5676cd2031b521e/hpc-test-node_W1) | :heavy_check_mark: SUCCESS
[hpc-test-node_W2](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/2235c1411ee4a6c15decdc20bdaf614e52fdde2c/hpc-test-node_W2) | :heavy_check_mark: SUCCESS
[hpc-test-node_W3](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/670080b5a2bbe06c96b52238ef95946973d03eda/hpc-test-node_W3) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/d41fb1af9b140a95e88060efa32aa5df981168e2/r-vmb-ppc-jenkins_W0) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/b2effccc27a2462b5918cee3a8e7f4b6233c6aaa/r-vmb-ppc-jenkins_W1) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/ed17dbe7c8296caa7a89e8264baf98e4a2abf4bd/r-vmb-ppc-jenkins_W2) | :heavy_check_mark: SUCCESS
[r-vmb-ppc-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/3875c16dc7e7035d2ce8ec7d206f2fff939fbb4a/r-vmb-ppc-jenkins_W3) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W0](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/010f6676296c8a6b4b4123c31bf78602d8865518/vulcan-jenkins_W0) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W1](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/485c885421a9b556082ed8964de0e53288002483/vulcan-jenkins_W1) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W2](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/d4e19a4ead56dfe238bda8a7820b762cc3e2b707/vulcan-jenkins_W2) | :heavy_check_mark: SUCCESS
[vulcan-jenkins_W3](https://gist.githubusercontent.com/mellanox-github/6121d76d7aa24f346ecc40b33e656fc6/raw/c92777aae392a4e9a788fe1ece2591218e1f6d8b/vulcan-jenkins_W3) | :heavy_check_mark: SUCCESS

File: src/ucp/tag/eager_snd.c [hunk 1]
```diff
@@ -253,10 +253,19 @@ static ucs_status_t ucp_tag_eager_sync_bcopy_multi(uct_pending_req_t *self)
...
+void ucp_tag_eager_sync_zcopy_completion(uct_completion_t *self,
+                                         ucs_status_t status)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t,
+                                          send.state.uct_comp);
     if (req->send.state.dt.offset == req->send.length) {
-        ucp_request_send_buffer_dereg(req); /* TODO register+lane change */
-        ucp_tag_eager_sync_completion(req, UCP_REQUEST_FLAG_LOCAL_COMPLETED,
-                                      status);
+        ucp_tag_eager_sync_zcopy_req_complete(req, status);
```
Comment (yosefe) [inline src/ucp/tag/eager_snd.c]: this fix seems weird. if uct_am_zcopy, returns UCS_OK, ucp_tag_eager_sync_zcopy_completion should never be called

Comment (dmitrygladkov) [reply src/ucp/tag/eager_snd.c]: yes, right
`ucp_tag_eager_sync_zcopy_req_complete` is called in this case

Comment (dmitrygladkov) [reply src/ucp/tag/eager_snd.c]: `ucp_tag_eager_sync_zcopy_completion` is called from `ucp_request_t::uct_completion_t::func` callback when Zcopy operation is completed

Comment (yosefe) [reply src/ucp/tag/eager_snd.c]: it shouldn't be called if uct_am_zcopy returns OK
maybe it's an issue with TCP implementation for zcopy

Comment (dmitrygladkov) [reply src/ucp/tag/eager_snd.c]: discussed offline with @yosefe 

by this:

> yes, right
> `ucp_tag_eager_sync_zcopy_req_complete` is called in this case

I meant that this is called from https://github.com/openucx/ucx/blob/1e227c972f1724d02751f65ba6196ea9d29a3349/src/ucp/proto/proto_am.inl#L197

Comment (yosefe) [reply src/ucp/tag/eager_snd.c]: yes, the problem is that complete() is called before dt.offset is updated, so it fails to complete UCP request.

**Pull request merged**